src/protocol/malicious.rs

```diff
@@ -0,0 +1,414 @@
+
+            v.accumulate_macs(
+                a_ctx.narrow(&Step::ValidateInput),
+                RecordId::from(0),
+                a_malicious,
+            );
+            v.accumulate_macs(
+                b_ctx.narrow(&Step::ValidateInput),
+                RecordId::from(1),
+                b_malicious,
+            );
+
+            let (ab, rab) = try_join(
+                a_ctx
+                    .narrow(&"SingleMult".to_string())
+                    .multiply(RecordId::from(0))
+                    .await
+                    .execute(a_shares[i], b_shares[i]),
+                a_ctx
+                    .narrow(&"DoubleMult".to_string())
+                    .multiply(RecordId::from(1))
+                    .await
+                    .execute(ra, b_shares[i]),
+            )
+            .await?;
+
+            v.accumulate_macs(
+                a_ctx.narrow(&Step::ValidateMultiplySubstep),
+                RecordId::from(0),
+                MaliciousReplicated::new(ab, rab),
+            );
+
+            v.validate(ctx.narrow(&"SecurityValidatorValidate".to_string()))
+                .await
+        });
+
+        try_join_all(futures).await?;
+
+        Ok(())
+    }
+
+    /// This is a big more complex arithmetic circuit that tests the validator a bit more thoroughly
+    /// input1   -
+    ///              input1 * input2
+    /// input2   -
+    ///              input2 * input3
+    /// input3   -
+    /// ...
+    /// input98  -
+    ///              input98 * input99
+    /// input99  -
+    ///              input99 * input100
+    /// input100 -
+    ///
+    /// This circuit has 100 inputs. Each input is multiplied with the adjacent inputs to produce 99 outputs.
+    ///
+    /// To acheive malicious security, the entire circuit must be run twice, once with the original inputs,
+    /// and once with all the inputs times a random, secret-shared value `r`. Two information theoretic MACs
+    /// are updated; once for each input, and once for each multiplication. At the end of the circuit, these
+    /// MACs are compared. If any helper deviated from the protocol, chances are that the MACs will not match up.
+    /// There is a small chance of failure which is `2 / |F|`, where `|F|` is the cardinality of the prime field.
+    #[tokio::test]
+    async fn test_complex_circuit() -> Result<(), BoxError> {
+        logging::setup();
+
+        let world: TestWorld = make_world(QueryId);
+        let context = make_contexts(&world);
+        let mut rng = rand::thread_rng();
+
+        let mut shared_inputs = [
+            Vec::with_capacity(100),
+            Vec::with_capacity(100),
+            Vec::with_capacity(100),
+        ];
+        for _ in 0..100 {
+            let x = Fp31::from(rng.gen::<u128>());
+            let x_shared = share(x, &mut rng);
+            shared_inputs[0].push(x_shared[0]);
+            shared_inputs[1].push(x_shared[1]);
+            shared_inputs[2].push(x_shared[2]);
+        }
```

Comment:
It's because down below I want to have vectors of JUST the shares belonging to a given helper. This code would produce a single array of these triples, which would be much harder to work with.